feat (build): Integrate protobuf using nanopb

ujjwal-cyph opened the pull request:
Integrate protobuf using nanopb into build system.

Fixes https://app.clickup.com/t/9002019994/CHI-3948

Due to a limitation of nanopb scripts (Ref: https://github.com/nanopb/nanopb/issues/527), it not possible to direct generated protob sources to desired directory via cmake configuration. So we cannot use cmake configuration of nanopb and a bash script is being used inside cmake.
We can look into this at a later stage.

rohanagarwal94 commented:
Task linked: [CHI-3948 Protobuf build integration for SDK](https://app.clickup.com/t/9002019994/CHI-3948)

kartikCypherock commented on utilities/cmake/simulator/simulator.cmake:
```diff
@@ -38,6 +38,9 @@ endif()
 target_include_directories(${PROJECT_NAME} PRIVATE
         src/
 
+        ${CMAKE_CURRENT_BINARY_DIR}
```
What is the need to add `${CMAKE_CURRENT_BINARY_DIR)` into the include path?

ujjwal-cyph replied:
the generated files are at the root of cmake build directory hence need to provide that for including headers.

ujjwal-cyph replied:
i tried `src/../` in place of `${CMAKE_CURRENT_BINARY_DIR}` but did not work. The generated protobuf headers could not be located.

ujjwal-cyph replied:
The example provided by nanopb (ref `vendor/nanopb/extra/FindNanopb.cmake#63`) mentions to use `CMAKE_CURRENT_BINARY_DIR` cmake variable for including protoc generated headers.

kartikCypherock commented:
I have one generic comment related to file deletions in `common/libraries/nanopb`, my understanding is that these files were required to support wallet connect related communication. I hope there is no impact expected on current features if we use generated pb decoders instead of a tested decoder file?

ujjwal-cyph commented:
> I have one generic comment related to file deletions in `common/libraries/nanopb`, my understanding is that these files were required to support wallet connect related communication. I hope there is no impact expected on current features if we use generated pb decoders instead of a tested decoder file?

Did a diff comparison of the files. Following differences noticed:
- nanopb `v0.4.6` is being updated to `v0.4.7` with this PR
  - no concerns IMO
- nanopb was being used with malloc
  - corrected this (added macro switch in the compile definition)

kartikCypherock commented:
> > I have one generic comment related to file deletions in `common/libraries/nanopb`, my understanding is that these files were required to support wallet connect related communication. I hope there is no impact expected on current features if we use generated pb decoders instead of a tested decoder file?
> 
> Did a diff comparison of the files. Following differences noticed:
> 
> * nanopb `v0.4.6` is being updated to `v0.4.7` with this PR
>   
>   * no concerns IMO
> * nanopb was being used with malloc
>   
>   * corrected this (added macro switch in the compile definition)

Ok, cool
Let's loop in @TejasvCypherock to confirm all compilation definitions set in this PR for nanopb match the same as done before

kartikCypherock commented on utilities/cmake/simulator/simulator.cmake:
```diff
@@ -20,7 +20,7 @@ set(EXECUTABLE ${PROJECT_NAME})
 include_directories(${SDL2_INCLUDE_DIRS})
 
-add_executable(${EXECUTABLE} ${SOURCES} ${CMAKE_CURRENT_BINARY_DIR}/version.c ${INCLUDES})
+add_executable(${EXECUTABLE} ${SOURCES} ${CMAKE_CURRENT_BINARY_DIR}/version.c ${COMMON_SRCS} ${COMMON_HDRS} ${PROTO_SRCS} ${PROTO_HDRS} ${INCLUDES})
```
```suggestion
add_executable(${EXECUTABLE} ${SOURCES} ${CMAKE_CURRENT_BINARY_DIR}/version.c ${PROTO_SRCS} ${PROTO_HDRS} ${INCLUDES})
```

kartikCypherock replied:
We don't need `$COMMON_SRCS` and `$COMMON_HDRS` anymore

ujjwal-cyph commented on CMakeLists.txt:
```diff
@@ -31,4 +52,12 @@ if("${CMAKE_BUILD_PLATFORM}" STREQUAL "Device")
+endif()
+
+# Include nanopb (at ${NANOPB_INCLUDE_DIRS}) and protoc-generated (at
+# ${CMAKE_CURRENT_BINARY_DIR}) headers
```
```suggestion
# ${CMAKE_CURRENT_BINARY_DIR}) headers. Ref: vendor/nanopb/extra/FindNanopb.cmake#63
```

amanCypherock commented:
We should also check all CI-related changes which are not triggered already.

ujjwal-cyph commented:
> We should also test all CI-related changes which are not triggered already.

Tested and fixed.

Reference run: https://github.com/ujjwal-cyph/x1_wallet_firmware/actions/runs/4828683503

TejasvCypherock reviewed (changes requested):
Need to remove old nanopb path from include in firmware.cmake

TejasvCypherock reviewed (commented):
We could also rename the `protob` submodule to its original name i.e. `cypherock-common`. As we might want to add other files that are not related to protobuf

TejasvCypherock commented on CMakeLists.txt:
```diff
@@ -15,6 +15,27 @@ endif()
+# btc.error.proto) as it would result into a collision.
+# By using <> format instead of "proto/" helps because in future, relocation
+# of generated files would not require updating this parameter.
+set(NANOPB_OPTIONS ",--generated-include-format=#include <%s>")
```
We might also want to consider adding a `--c-style` flag to better match our existing coding style

ujjwal-cyph replied:
added

TejasvCypherock reviewed (changes requested):
As the files are directly generated in the build folder, It is not clear to me how will we be using these files while programming as the IDE would need its reference for intellisense and other features. You might want to set up a simple example to clarify this situation. Also the `--c-style` flag doesn't seem to work.

ujjwal-cyph commented:
> As the files are directly generated in the build folder, It is not clear to me how will we be using these files while programming as the IDE would need its reference for intellisense and other features. You might want to set up a simple example to clarify this situation. Also the `--c-style` flag doesn't seem to work.

- Fixed the issue with `--c-style` parameter (https://github.com/Cypherock/x1_wallet_firmware/pull/222/commits/4eb09cc109839b0e462e03c0c84495e065132653).
- Added a bash script to build protob sources for IDE support (https://github.com/Cypherock/x1_wallet_firmware/pull/222/commits/e7704807f1e6bff751562346b12f34c9f3d05827).
